pkg/build/builder/common.go

```diff
@@ -448,7 +448,7 @@ func replaceImagesFromSource(node *parser.Node, imageSources []buildapiv1.ImageS
 		}
 	}
 	names := make(map[string]string)
-	stages, err := imagebuilder.NewStages(node, imagebuilder.NewBuilder(nil))
+	stages, err := imagebuilder.NewStages(node, imagebuilder.NewBuilder(make(map[string]string)))
```

Comment:
the prior code only threw an error on
```
ARG foo=bar
FROM centos
```

it did not throw an error on:
```
FROM centos
ARG foo=bar
```

so i'd like to better understand why this is the right fix before we merge it.

Reply:
also, we normally call both `findReferencesImages` and `replaceImagesFromSource` (in that order).

it's the second call that fails.

but both functions invoke:
` imagebuilder.NewStages(node, imagebuilder.NewBuilder(make(map[string]string)))`

so why does the second invocation fail and not the first?  Some stateful mutation of `node`?

Reply:
So the panic occurred at https://github.com/openshift/builder/blob/master/vendor/github.com/openshift/imagebuilder/dispatchers.go#L557

With `arg` method, the `Builder` object ptr passed in corresponds to the `imagebuilder.NewBuilder` method call here, where the map gets set to the `Args` field in the `Builder` object.

The call to `NewStages` leads to a call on this function (which is a function on the `Builder` created by `imagebuild.NewBuilder`):

```
func (b *Builder) extractHeadingArgsFromNode(node *parser.Node) error {
	var args []*parser.Node
	var children []*parser.Node
	extract := true
	for _, child := range node.Children {
		if extract && child.Value == command.Arg {
			args = append(args, child)
		} else {
			if child.Value == command.From {
				extract = false
			}
			children = append(children, child)
		}
	}

	for _, c := range args {
		step := b.Step()
		if err := step.Resolve(c); err != nil {
			return err
		}
		if err := b.Run(step, NoopExecutor, false); err != nil {
			return err
		}
	}

	node.Children = children
	return nil
}
```

The `step` in that final for loop corresponds to the `ARG` step in the Dockerfile.  And the `b.Run` call results in a call to `func arg(....)`.

I interpreted the need for having to supply the map for `imagebuilder.NewBuilder`, vs. that method creating the map, as analogous to `--build-args`with the various k/v pairs associated in the various ARGS steps.  If you had those, they would be in the map passed in.

At this, we are only filling in ARGS with have default values.  So the empty map suffices.

It did not throw the exception on the later case since the ARG is defined after the FROM, it cannot be used with the FROM.

I *think* that covers it.

Reply:
https://github.com/openshift/imagebuilder/pull/147#issue-373860798 is where I'd like to fix the panic in imagebuilder, but I expect that this patch here would also prevent it.

Reply:
it would .... and we could always do this short term, and then switch it after an imagebuilder merge and vendor bump here.

But longer term @nalind , we still need to reconcile in the slack thread between you, @bparees , and I the need to do the default value ARG substitution for FROMs in openshift/builder via imagebuilder vs. having imagebuildah.BuildDockerfile do it somehow.

Reply:
> also, we normally call both findReferencesImages and replaceImagesFromSource (in that order).

i think i was wrong about this, sorry for the noise.

Reply:
RE: the [2nd slack discussion thread](https://coreos.slack.com/archives/CE2HALN2W/p1581451537394100) @bparees gave @nalind a simplified reproducer which bypasses the panic fixed with this PRs changes and produces the issue with ARG defaults not showing up in FROM references:

```
apiVersion: build.openshift.io/v1
kind: BuildConfig
metadata:
  name: sample-build-docker-args-preset
spec:
  source:
    dockerfile: |-
      FROM centos
      ARG foo=centos
      FROM $foo
    type: Dockerfile
  strategy:
    dockerStrategy: {}
    type: Docker
```

which results in:
```
I0211 22:06:43.140914       1 builder.go:318] Starting Docker build from build config sample-build-docker-args-preset-17 ...
F0211 22:06:43.209605       1 helpers.go:114] error: build error: error parsing name "$foo": error parsing image name "$foo": invalid reference format
```

@nalind said he would start poking around at it.

I'll update the cloned BZ on the buildah queue with this.